crates/procedurals/src/lib.rs

```diff
@@ -0,0 +1,178 @@
+                        let mesh_aabb = mesh.aabb();
+                        let gpu_mesh = GpuMesh::from_mesh(&assets, mesh);
+                        (gpu_mesh, mesh_aabb)
+                    };
+                    world
+                        .add_components(
+                            id,
+                            Entity::new()
+                                .with(ambient_core::mesh(), gpu_mesh)
+                                .with_default(main_scene())
+                                .with_default(gpu_primitives_mesh())
+                                .with_default(gpu_primitives_lod())
+                                .with_default(primitives())
+                                .with_default(local_to_world())
+                                .with_default(mesh_to_world())
+                                .with(local_bounding_aabb(), mesh_aabb)
+                                .with(world_bounding_aabb(), mesh_aabb)
+                                .with(world_bounding_sphere(), mesh_aabb.to_sphere()),
+                        )
+                        .unwrap();
+                }
+            }),
+            query(procedural_material().changed()).to_system(|query, world, query_state, _| {
+                let assets = world.resource(asset_cache()).clone();
+                for (id, material_handle) in query.collect_cloned(world, query_state) {
+                    let storage = world.resource(procedural_storage());
+                    let material = storage.get_material(material_handle).clone();
+                    let material = PbrMaterial::new(&assets, material);
+                    let material = SharedMaterial::new(material);
+                    world
+                        .add_components(
+                            id,
+                            Entity::new()
+                                .with(ambient_renderer::material(), material)
+                                .with(renderer_shader(), cb(get_pbr_shader)),
+                        )
+                        .unwrap();
+                }
+            }),
+        ],
+    )
+}
+
+pub type ProceduralMesh = Mesh;
+pub type ProceduralTexture = Arc<TextureView>;
+pub type ProceduralSampler = Arc<wgpu::Sampler>;
+pub type ProceduralMaterial = PbrMaterialConfig;
+
+#[derive(Clone)]
+pub struct ProceduralStorage {
+    meshes: HashMap<ProceduralMeshHandle, ProceduralMesh>,
+    textures: HashMap<ProceduralTextureHandle, ProceduralTexture>,
+    samplers: HashMap<ProceduralSamplerHandle, ProceduralSampler>,
+    materials: HashMap<ProceduralMaterialHandle, ProceduralMaterial>,
+}
+
+impl ProceduralStorage {
+    pub fn new() -> Self {
+        Self {
+            meshes: Default::default(),
+            textures: Default::default(),
+            samplers: Default::default(),
+            materials: Default::default(),
+        }
+    }
+
+    pub fn insert_mesh(&mut self, mesh: ProceduralMesh) -> ProceduralMeshHandle {
```

Comment:
All of these methods are just redirecting to the underlying values; you can just make them public instead and save on the boilerplate

Reply:
I would like to keep these for now, since we might add more things in the near future, which might require denying free access to the internal state. However if that doesn't happen I might refactor them later.

Reply:
I'd suggest renaming this to `AllProceduralStorage` (or similar) creating a `ProceduralStorage<ProceduralMaterialHandle, ProceduralMaterial>` struct and making those public, then blocking off access to the internal state, so that you can do `all.textures.get(...)` and ensure they go through the same API without getting direct access and without having to define getters/setters for each resource type.

I think this machinery could also be used for other kinds of procedurals, like #182, so I'd like to reduce the effort required to add a new procedural in future.